src/tamp_improv/approaches/improvisational/policies/goal_rl.py

```diff
@@ -0,0 +1,377 @@
+        """Get the success rate over all training."""
+        if not self.success_history:
+            return 0.0
+        return float(sum(self.success_history) / len(self.success_history))
+
+    @property
+    def _get_avg_episode_length(self) -> float:
+        """Get the average episode length over all training."""
+        if not self.episode_lengths:
+            return 0.0
+        return float(np.mean(self.episode_lengths))
+
+    @property
+    def _get_avg_reward(self) -> float:
+        """Get the average reward over all training."""
+        if not self.episode_rewards:
+            return 0.0
+        return float(np.mean(self.episode_rewards))
+
+
+class GoalConditionedRLPolicy(Policy[ObsType, ActType]):
+    """Goal-conditioned RL policy for learning shortcuts."""
+
+    def __init__(self, seed: int, config: GoalConditionedRLConfig | None = None):
+        """Initialize policy."""
+        super().__init__(seed)
+        self.config = config or GoalConditionedRLConfig()
+        if self.config.device == "cuda":
+            self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        else:
+            self.device = self.config.device
+        self.model: SAC | TD3 | None = None
+        self.node_states: dict[int, ObsType] = {}
+        self.valid_shortcuts: list[tuple[int, int]] = []
+        self._current_context: PolicyContext | None = None
+        self._current_goal: ObsType | None = None
+
+    @property
+    def requires_training(self) -> bool:
+        """Whether this policy requires training data and training."""
+        return True
+
+    def initialize(self, env):
+        """Initialize policy with environment."""
+
+    def can_initiate(self) -> bool:
+        """Check whether the policy can be executed."""
+        return self.model is not None and self._current_goal is not None
+
+    def configure_context(self, context: PolicyContext) -> None:
+        """Configure policy with context information."""
+        self._current_context = context
+        target_node_id = context.info.get("target_node_id")
+        assert target_node_id is not None
+        self._current_goal = self.node_states[target_node_id]
+
+    def get_action(self, obs: dict[str, np.ndarray[Any, Any]]) -> np.ndarray[Any, Any]:  # type: ignore[override] # pylint: disable=line-too-long
+        """Get action from policy."""
+        assert self.model is not None and self._current_goal is not None
+        assert isinstance(
+            obs, dict
+        ), "Observation must be a dictionary, consistent with HER"
+        action, _ = self.model.predict(obs, deterministic=True)
+        return action
+
+    def train(
+        self,
+        env: gym.Env,
+        train_data: GoalConditionedTrainingData,  # type: ignore[override]
+    ):
+        """Train the policy."""
+        print(f"\nTraining goal-conditioned RL policy ({self.config.algorithm})...")
+
+        node_states = getattr(train_data, "node_states", None)
```

Comment:
I want to encourage you to rely less on `getattr` and `hasattr` in your code generally. it's more "pythonic" to assume that attributes either exist or don't exist (and if that assumption is wrong, the code should crash). in this case, is it possible to just do `self.node_states = train_data.node_states` and `self.valid_shortcuts = train_data.valid_shortcuts`?